Run the E2E tests as a user with the author role

notnownikki opened the pull request:
## Description

Adds support for `E2E_ROLE` env var which allows the e2e tests to run as a user with the `author` role.

It adds a dummy `favicon.ico` so that we don't (sometimes) error on the 404 when looking at front end posts, and ignores ` net::ERR_UNKNOWN_URL_SCHEME` console message which we also (sometimes) get when viewing a front end post.

The code that deals with plugins switches to the admin user, and back to the test user after the plugin operations are done.

## How has this been tested?

Both e2e tests should pass.

At the moment, they require https://github.com/WordPress/gutenberg/pull/11395 because the fail is catching that bug.

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

notnownikki committed:
Run the E2E tests as a user with the author role

notnownikki commented on test/e2e/specs/change-detection.test.js:
```diff
@@ -94,7 +94,7 @@ describe( 'Change detection', () => {
 		await ensureSidebarOpened();
 
-		const postStickyToggleButton = ( await page.$x( "//label[contains(text(), 'Stick to the Front Page')]" ) )[ 0 ];
+		const postStickyToggleButton = ( await page.$x( "//label[contains(text(), 'Pending Review')]" ) )[ 0 ];
```
Authors can mark things needing review, they can't stick things to the front page. The effect in the test is the same.

aduth replied:
If I'm reading correctly, the E2E tests can still be run by a non-author, in which case the label would not have the changed effect here. Should this be made more generic? Lookup element by its class? A helper utility which returns the button accounting for role?

notnownikki replied:
As far as I can see, no matter what your user role, clicking "Pending Review" marks the post as dirty, and that's what's getting tested here.

aduth replied:
Oh, I think I misunderstood the reason for the change. It’s just changing it to an option which is visible for both admins and authors.

notnownikki commented on test/e2e/specs/templates.test.js:
```diff
@@ -60,12 +62,12 @@ describe( 'templates', () => {
 		const STANDARD_FORMAT_VALUE = '0';
 
 		async function setPostFormat( format ) {
+			await switchToAdminUser();
```
Switching user will not actually do anything if we're running as the admin user. We'll only switch if we're running as a different user, because the author role cannot manage plugins.

tofumatt replied:
That's good to know; could you leave a comment explaining that in the code? 😄

notnownikki commented on test/e2e/support/setup-test-framework.js:
```diff
@@ -116,6 +116,12 @@ function observeConsoleLogging() {
 
+		// Viewing posts on the front end can result in this error, which
+		// has nothing to do with Gutenberg.
+		if ( text.includes( 'net::ERR_UNKNOWN_URL_SCHEME' ) ) {
```
This seems to be a jQuery thing that happens on the front end, and we sometimes would get hit by it, but sometimes the test assertions would complete quickly enough for it not to make the test fail.

tofumatt replied:
Is there an issue anywhere that tracks that? I don't quite follow.

notnownikki replied:
I'm not sure what causes it yet, it seems to be unrelated to Gutenberg, and happens when you view the published post on the front end. I'll investigate further after this merge.

aduth commented on test/e2e/specs/templates.test.js:
```diff
@@ -60,12 +62,12 @@ describe( 'templates', () => {
+			await switchToAdminUser();
 			await visitAdmin( 'options-writing.php' );
 			await page.select( '#default_post_format', format );
-			return Promise.all( [
```
This was accounting for a race condition acknowledged in the Puppeteer docs:

>Bear in mind that if click() triggers a navigation event and there's a separate page.waitForNavigation() promise to be resolved, you may end up with a race condition that yields unexpected results. The correct pattern for click and wait for navigation is the following:

https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pageclickselector-options

aduth commented on test/e2e/specs/change-detection.test.js:
```diff
@@ -94,8 +94,8 @@ describe( 'Change detection', () => {
 		// Toggle post as sticky (not persisted for autosave).
```
We're no longer toggling as sticky.

notnownikki commented:
Not sure what's going on here, I've merged master but the test still fails for authors when it passes on my local branch.

notnownikki commented:
So this is all working correctly now, and is catching an actual bug.

To recreate:

Set up a brand new WP install.
Set up an author user.
Create a new post, publish it.
Create a new post and add a "Latest Posts" block.
See the 403 error in the console.

Seems that the Latest Posts block is doing something with category information, and requesting `/index.php?rest_route=%2Fwp%2Fv2%2Fcategories&per_page=100&context=edit&_locale=user` which an author cannot access.

cc: @youknowriad @danielbachhuber

notnownikki committed:
Fix latest posts block for authors

notnownikki commented:
@aduth I think I addressed your points here, and this would be good to get in because it caught a real bug.

@youknowriad Change to the Latest Posts block is included here now.

danielbachhuber commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -27,15 +29,33 @@ import {
 
+	componentWillMount() {
+		apiFetch( {
+			path: addQueryArgs( `/wp/v2/categories`, CATEGORIES_LIST_QUERY ),
```
Doesn't this need to iterate the entire list? What if I have 102 categories?

notnownikki replied:
:woman_shrugging: this is just how it was in the original code, I was just fixing it so that it didn't break for authors

danielbachhuber replied:
Ok. I'll mention on #10873 and we can resolve with that.

youknowriad commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -27,15 +29,33 @@ import {
+			path: addQueryArgs( `/wp/v2/categories`, CATEGORIES_LIST_QUERY ),
+		} ).then(
+			( categoriesList ) => {
+				this.setState( { categoriesList } );
```
What if the block is removed at this point, we should guard against it

notnownikki replied:
Added an abort call when the component unmounts, same as the hierarchical term selector.

notnownikki committed:
About fetch when unmounting

youknowriad commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -56,6 +56,10 @@ class LatestPostsEdit extends Component {
 	}
 
+	componentWillUnmount() {
+		invoke( this.fetchRequest, [ 'abort' ] );
```
Do we support "abort" in apiFetch requests?

notnownikki replied:
if we don't, then the hierarchical term selector is wrong too.

youknowriad replied:
I don't think we do :). I think we did at some point in apiRequest because we were relying on jQuery but it's not the case anymore AFAIK.

notnownikki replied:
Yup, no abort! Ok, I'll do the `this.isStillMounted` thing instead. Can we fix the term selector in another PR? I'm keen to get this in with as few changes as possible so authors get tested asap

notnownikki committed:
Don't set state if we're not still mounted

notnownikki committed:
Update run-e2e-tests.sh

tofumatt commented on test/e2e/specs/templates.test.js:
```diff
@@ -93,9 +97,13 @@ describe( 'templates', () => {
 		} );
 
 		it( 'should not populate new page with default block for format', async () => {
+			// This test always needs to run as the admin user.
```
Could you explain why it needs to run as the admin user?

tofumatt committed:
chore: Tweak function docs

tofumatt reviewed (approved):
This looks good to me and is a solid improvement.

<del>I left a few comments and I'm not sure if other comments are resolved (they aren't marked as such automatically by GitHub, so I've flagged @aduth for another review).</del> Never mind, I think maybe I had this tab left open and missed the feedback being addressed. :shipit: 

I think this is a solid improvement though, cheers!

notnownikki commented on .travis.yml:
```diff
@@ -67,5 +67,9 @@ jobs:
     - stage: test
       env: WP_VERSION=latest
       script:
-        - npm install || exit 1
```
It's run as part of the e2e tests. It didn't used to be, but now it is.

notnownikki committed:
If fetching categories fails, default to empty list

notnownikki commented on packages/block-library/src/latest-posts/edit.js:
```diff
@@ -27,15 +29,40 @@ import {
+		this.isStillMounted = true;
+		this.fetchRequest = apiFetch( {
+			path: addQueryArgs( `/wp/v2/categories`, CATEGORIES_LIST_QUERY ),
+		} ).then(
```
Added a `catch` which sets the list to an empty list.

notnownikki committed:
Comment update
